### PR TITLE
 JAMES-2767 Replace old ES modules by the new one

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

**Comment by chibenwa** [comment]
Have you checked `server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ElasticSearchQuotaSearchExtension.java`

I bet there would be an issue there

**Comment by chibenwa** [comment]
Looks awesome!

**Comment by chibenwa** [comment]
```
[a3f3019e3b627d2221df8f79287c9d831523e1b8] [ERROR]   ElasticSearchListeningMessageSearchIndexTest.deserializeElasticSearchListeningMessageSearchIndexGroup:101 » GroupDeserialization
```

**Comment by Arsnael** [comment]
Read it ! 

I'm not sure now that https://github.com/linagora/james-project/pull/2387 is still necessary as you seem to do pretty much the same work in this commit https://github.com/linagora/james-project/pull/2388/commits/5abf9b9398a9fdfe82be7f744524a52ee857d2e4

**Comment by trantienduchn** [comment]
> I'm not sure now that #2387 is still necessary as you seem to do pretty much the same work in this commit [5abf9b9](https://github.com/linagora/james-project/commit/5abf9b9398a9fdfe82be7f744524a52ee857d2e4)

yeah I think the MPT migration is just about updating some pieces of code in ElastichSearchHostSystem

**Comment by Arsnael** [comment]
> > I'm not sure now that #2387 is still necessary as you seem to do pretty much the same work in this commit [5abf9b9](https://github.com/linagora/james-project/commit/5abf9b9398a9fdfe82be7f744524a52ee857d2e4)
> 
> yeah I think the MPT migration is just about updating some pieces of code in ElastichSearchHostSystem

It's like this exactly

**Comment by trantienduchn** [comment]
```
[e2d8fa1f764d36493675a8625918abf6703bc4f8] 09:07:45.051 [31m[WARN ][0;39m o.a.j.m.m.ElasticSearchMailboxModule - Error establishing ElasticSearch connection. Next retry scheduled in PT3S ms
[e2d8fa1f764d36493675a8625918abf6703bc4f8] org.apache.http.ParseException: Not a valid protocol version: This is not a HTTP port
```
suspecting this is a configuration error when running the validation tests.

**Comment by chibenwa** [comment]
` Next retry scheduled in PT3S ms`

-> Can we add a commit for better printing this ? <3

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java [hunk 1]
```diff
@@ -19,140 +19,177 @@
...
-        try {
-            createIndexIfNeeded(client, indexName, generateSetting(nbShards, nbReplica));
-            aliases.forEach(alias -> createAliasIfNeeded(client, indexName, alias));
-        } catch (IOException e) {
-            LOGGER.error("Error while creating index : ", e);
+        public RestHighLevelClient createIndexAndAliases(RestHighLevelClient client) {
+            Preconditions.checkNotNull(indexName);
+            try {
+                createIndexIfNeeded(client, indexName, generateSetting(nbShards, nbReplica));
+                aliases.forEach(Throwing.consumer(alias -> createAliasIfNeeded(client, indexName, alias)));
```
Comment (blackheaven) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java]: no sneakyThrow?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrollIterable.java [hunk 2]
```diff
@@ -20,56 +20,67 @@
...
+        return new ScrollIterator(client, searchRequest);
     }
 
     public Stream<SearchResponse> stream() {
         return Iterators.toStream(iterator());
     }
 
     public static class ScrollIterator implements Iterator<SearchResponse> {
+        private final RestHighLevelClient client;
+        private CompletableFuture<SearchResponse> searchResponseFuture;
```
Comment (blackheaven) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrollIterable.java]: Why not using Reactor here?

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java [hunk 3]
```diff
@@ -144,15 +143,16 @@ public void deleteByQueryShouldWorkOnSingleMessage() throws Exception {
...
         elasticSearch.awaitForElasticSearch();
         
-        try (Client client = getESClient()) {
-            SearchResponse searchResponse = client.prepareSearch(INDEX_NAME.getValue())
-                    .setTypes(TYPE_NAME.getValue())
-                    .setQuery(QueryBuilders.matchAllQuery())
-                    .get();
-            assertThat(searchResponse.getHits().getTotalHits()).isEqualTo(0);
+        try (RestHighLevelClient client = getESClient()) {
+            await().atMost(Duration.TEN_SECONDS)
```
Comment (blackheaven) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: use any CALMLY_AWAIT please

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java [hunk 4]
```diff
@@ -174,20 +174,21 @@ public void deleteByQueryShouldWorkWhenMultipleMessages() throws Exception {
...
         elasticSearch.awaitForElasticSearch();
         
-        try (Client client = getESClient()) {
-            SearchResponse searchResponse = client.prepareSearch(INDEX_NAME.getValue())
-                    .setTypes(TYPE_NAME.getValue())
-                    .setQuery(QueryBuilders.matchAllQuery())
-                    .get();
-            assertThat(searchResponse.getHits().getTotalHits()).isEqualTo(1);
+        try (RestHighLevelClient client = getESClient()) {
+            await().atMost(Duration.TEN_SECONDS)
```
Comment (blackheaven) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: idem

**Comment by Arsnael** [comment]
Passed the tests, crashed at the packaging after

test this please

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrollIterable.java (hunk 2 above)
Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrollIterable.java]: I think using Mono here is verbose, the CompletableFuture is returned by ElasticSearch client response (so anyway we cannot avoid that), and the future's features is enough for the implementation.

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java (hunk 3 above)
Comment (trantienduchn) [reply backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: I can't find any CALMLY_AWAIT that I can import (I don't want to import non related dependencies), so I created a new one

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java (hunk 1 above)
Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java]: check it out 94f9f7a

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java (hunk 3 above)
Comment (trantienduchn) [reply backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: check it out 94f9f7a

**Comment by trantienduchn** [comment]
```
...
[94f9f7a990c428f85c8c72b5730adf0ad546ecbf] 	Suppressed: org.elasticsearch.client.ResponseException: method [PUT], host [http://elasticsearch:9200], URI [/mailbox_v1?master_timeout=30s&include_type_name=false&timeout=30s], status line [HTTP/1.1 400 Bad Request]
[94f9f7a990c428f85c8c72b5730adf0ad546ecbf] {"error":{"root_cause":[{"type":"index_already_exists_exception","reason":"already exists","index":"mailbox_v1"}],"type":"index_already_exists_exception","reason":"already exists","index":"mailbox_v1"},"status":400}
...
```
passed mvn test, this result is expected, CI is still using ES 2.

**Comment by chibenwa** [comment]
```
[INFO] Apache James :: Server :: JMAP :: Cassandra Integration testing SUCCESS [06:03 min]
```

Was:

```
[INFO] Apache James :: Server :: JMAP :: Cassandra Integration testing SUCCESS [05:47 min]
```

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java [hunk 5]
```diff
@@ -151,10 +154,14 @@ public void run() {
...
 
     private boolean checkMetricRecordedInElasticSearch() {
-        try (Client client = elasticSearchExtension.getDockerES().clientProvider().get()) {
-            return !Arrays.stream(client.prepareSearch()
-                    .setQuery(QueryBuilders.matchAllQuery())
-                    .get().getHits().getHits())
+        try (RestHighLevelClient client = elasticSearchExtension.getDockerES().clientProvider().get()) {
+            SearchRequest searchRequest = new SearchRequest()
+                .source(new SearchSourceBuilder()
+                .query(QueryBuilders.matchAllQuery()));
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java]: indent

**Comment by chibenwa** [comment]
```
[8cdf44790b4ee99fe91fcc8ab43a9995f4533ab5] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
[8cdf44790b4ee99fe91fcc8ab43a9995f4533ab5] [ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:475 » Execution
```

test this please

**Comment by trantienduchn** [comment]
```
[8cdf44790b4ee99fe91fcc8ab43a9995f4533ab5] 	Wrapped by: org.elasticsearch.client.ResponseException: method [PUT], host [http://elasticsearch:9200], URI [/mailbox_v1/_mapping?master_timeout=30s&include_type_name=false&timeout=30s], status line [HTTP/1.1 400 Bad Request]
[8cdf44790b4ee99fe91fcc8ab43a9995f4533ab5] {"error":{"root_cause":[{"type":"action_request_validation_exception","reason":"Validation Failed: 1: mapping type is missing;"}],"type":"action_request_validation_exception","reason":"Validation Failed: 1: mapping type is missing;"},"status":400}
```
this is the case of running new james with on old ES server, seems the log messages is not quite easy to understand what just happend

**Comment by chibenwa** [comment]
> this is the case of running new james with on old ES server, seems the log messages is not quite easy to understand what just happend

@trantienduchn can you write the **StartUpCheck** for ES 6 in order to make this more clear?

File: mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/json/QuotaRatioToElasticSearchJsonTest.java [hunk 6]
```diff
@@ -69,9 +69,8 @@ void quotaRatioShouldBeWellConvertedToJsonWhenNoDomain() throws IOException {
             .instant(Instant.now())
             .build();
 
-
         QuotaRatioToElasticSearchJson quotaRatioToElasticSearchJson = new QuotaRatioToElasticSearchJson();
-        String convertToJson = quotaRatioToElasticSearchJson.convertToJson(user, event);
+        String convertToJson = quotaRatioToElasticSearchJson.convertToJson( event);
```
Comment (blackheaven) [inline mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/json/QuotaRatioToElasticSearchJsonTest.java]: extra space

**Comment by chibenwa** [comment]
Somehow :green_apple: before CI restart

**Comment by chibenwa** [comment]
Merged

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**